internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -46,75 +37,63 @@ func (k *Kubeadm) ID(data *userdata.UserData) string {
 // PreFunc implements the Service interface.
 // nolint: gocyclo
 func (k *Kubeadm) PreFunc(data *userdata.UserData) (err error) {
-	if err = writeKubeadmConfig(data); err != nil {
+	if err = kubeadm.WriteConfig(data); err != nil {
 		return err
 	}
 
-	if data.Services.Kubeadm.IsBootstrap() {
-		// Write out all certs we've been provided
-		certs := []struct {
-			Cert     *x509.PEMEncodedCertificateAndKey
-			CertPath string
-			KeyPath  string
-		}{
-			{
-				Cert:     data.Security.Kubernetes.CA,
-				CertPath: constants.KubeadmCACert,
-				KeyPath:  constants.KubeadmCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.SA,
-				CertPath: constants.KubeadmSACert,
-				KeyPath:  constants.KubeadmSAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.FrontProxy,
-				CertPath: constants.KubeadmFrontProxyCACert,
-				KeyPath:  constants.KubeadmFrontProxyCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.Etcd,
-				CertPath: constants.KubeadmEtcdCACert,
-				KeyPath:  constants.KubeadmEtcdCAKey,
-			},
-		}
-
-		for _, cert := range certs {
-			if cert.Cert == nil {
-				continue
-			}
-			if err = writeKubeadmPKIFiles(cert.Cert, cert.CertPath, cert.KeyPath); err != nil {
-				return err
-			}
-		}
-	} else if data.Services.Kubeadm.IsControlPlane() {
-		if data.Services.Trustd == nil || data.Services.Trustd.BootstrapNode == "" {
-			return nil
+	// Write out all certs we've been provided
+	if data.Services.Kubeadm.IsControlPlane() {
+		if err = kubeadm.WritePKIFiles(data); err != nil {
+			return err
 		}
+	}
 
-		creds, err := basic.NewCredentials(data.Services.Trustd)
-		if err != nil {
+	// Run kubeadm init phase certs all. This should fill in whatever gaps
+	// we have in the provided certs.
+	if data.Services.Kubeadm.IsBootstrap() {
+		if err = kubeadm.PhaseCerts(); err != nil {
 			return err
 		}
+	}
 
-		files := []string{
-			constants.AuditPolicyPathInitramfs,
-			constants.EncryptionConfigInitramfsPath,
-		}
+	if data.Services.Kubeadm.IsWorker() || data.Services.Kubeadm.IsBootstrap() || data.Services.Trustd == nil {
+		log.Println("Skipping retrieval of files from peers via trustd")
+		return nil
+	}
 
-		conn, err := basic.NewConnection(data.Services.Trustd.BootstrapNode, constants.TrustdPort, creds)
-		if err != nil {
-			return err
+	// Initialize trustd peer client connection
+	var trustds []proto.TrustdClient
+	if trustds, err = kubeadm.CreateTrustdClients(data); err != nil {
+		return err
+	}
+
+	// Generate a list of files we need to request
+	// ( filtered by ones we already have ) and
+	// Get assets from remote nodes
+	for _, fileRequest := range kubeadm.FileSet() {
+		// Have a single chan shared across all clients
+		// for a given file
+		content := make(chan []byte)
+
+		ctx, ctxCancel := context.WithCancel(context.Background())
```

Comment:
it's a good habbit (and some linters might enforce it), to do `defer ctxCancel()` next line so that no matter what context resources do not leak. as this is a for loop, probably extracting loop body into a function might make sense

2nd thing is that we can add `ctx` parameter to the `PreFunc` as parameter, so that we can make this context be a child to that context. ServiceRunner has a context it can pass down, so that if service is shut down (e.g. on node shut down), it will abort `PreFunc` and this download